File: src/scraper/list.js
const axios = require('axios');
const cheerio = require('cheerio');
const { BASE_URL, USER_AGENT, REQUEST_TIMEOUT } = require('../config/config');
const { getRandomUserAgent } = require('../utils/helpers');

/**
 * Scrape the overall anime list from the website
 * @returns {Promise<Array>} Array of anime groups organized by first letter
 */
const scrapeAnimeList = async () => {
  try {
    // Set headers to mimic a browser request
    const headers = {
      'User-Agent': getRandomUserAgent() || USER_AGENT
    };
    
    // URL to scrape
    const url = `${BASE_URL}/anime-list/`;
    
    // Send HTTP request
    const response = await axios.get(url, {
      headers,
      timeout: REQUEST_TIMEOUT
    });
    
    // Parse HTML content
    const $ = cheerio.load(response.data);
    
    // Find anime list structure - typically organized by first letter
    const animeGroups = [];
    
    // Find all divs with class "letter-group" or similar
    const letterGroups = $(['div[class*="letter-group"]', 'ul.lcp_catlist']);
    
    // If the standard structure isn't found, try to find all list items
    if (letterGroups.length === 0) {
      const allListItems = $('li');
      const animeList = [];
      
      allListItems.each((index, item) => {
        const link = $(item).find('a');
        if (link.length) {
          const title = link.text().trim();
          const url = link.attr('href');
          
          animeList.push({
            title,
            url
          });
        }
      });
      
      animeGroups.push({
        letter: 'All',
        anime_list: animeList
      });
    } else {
      // Process each letter group
      letterGroups.each((index, group) => {
        // Try to find the letter/heading
        const heading = $(group).find(['h2', 'h3']);
        const letter = heading.length ? heading.text().trim() : "Unknown";
        
        // Find all anime links in this group
        const links = $(group).find('a');
        const animeList = [];
        
        links.each((i, link) => {
          const title = $(link).text().trim();
          const url = $(link).attr('href');
          
          animeList.push({
            title,
            url
          });
        });
        
        if (animeList.length) {
          animeGroups.push({
            letter,
            anime_list: animeList
          });
        }
      });
    }
    
    return animeGroups;
    
  } catch (error) {
    console.error(`Error fetching the anime list page: ${error.message}`);
    throw new Error(`Failed to scrape anime list: ${error.message}`);
  }
};

module.exports = {
  scrapeAnimeList
};

File: scraper/list.go
package scraper

import (
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/Pendetot/AnimekApi/utils"
)

// AnimeItem represents a single anime item
type AnimeItem struct {
	Title string `json:"title"`
	URL   string `json:"url"`
}

// AnimeGroup represents a group of anime organized by letter
type AnimeGroup struct {
	Letter    string      `json:"letter"`
	AnimeList []AnimeItem `json:"anime_list"`
}

// ScrapeAnimeList scrapes the overall anime list from the website
func ScrapeAnimeList() ([]AnimeGroup, error) {
	// Create HTTP client with timeout
	client := &http.Client{
		Timeout: 10 * time.Second,
	}

	// URL to scrape
	url := "https://ww1.anoboy.app/anime-list/"

	// Create request with headers
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %v", err)
	}

	req.Header.Set("User-Agent", utils.GetUserAgent())

	// Send request
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch page: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("received status code: %d", resp.StatusCode)
	}

	// Parse HTML
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to parse HTML: %v", err)
	}

	var animeGroups []AnimeGroup

	// Find anime list structure - typically organized by first letter
	letterGroups := doc.Find("div[class*="letter-group"], ul.lcp_catlist")

	// If the standard structure isn't found, try to find all list items
	if letterGroups.Length() == 0 {
		var animeList []AnimeItem
		
		allListItems := doc.Find("li")
		allListItems.Each(func(i int, item *goquery.Selection) {
			link := item.Find("a").First()
			if link.Length() > 0 {
				title := strings.TrimSpace(link.Text())
				if url, exists := link.Attr("href"); exists && title != "" {
					animeList = append(animeList, AnimeItem{
						Title: title,
						URL:   url,
					})
				}
			}
		})

		if len(animeList) > 0 {
			animeGroups = append(animeGroups, AnimeGroup{
				Letter:    "All",
				AnimeList: animeList,
			})
		}
	} else {
		// Process each letter group
		letterGroups.Each(func(i int, group *goquery.Selection) {
			// Try to find the letter/heading
			heading := group.Find("h2, h3").First()
			letter := "Unknown"
			if heading.Length() > 0 {
				letter = strings.TrimSpace(heading.Text())
			}

			// Find all anime links in this group
			links := group.Find("a")
			var animeList []AnimeItem

			links.Each(func(j int, link *goquery.Selection) {
				title := strings.TrimSpace(link.Text())
				if url, exists := link.Attr("href"); exists && title != "" {
					animeList = append(animeList, AnimeItem{
						Title: title,
						URL:   url,
					})
				}
			})

			if len(animeList) > 0 {
				animeGroups = append(animeGroups, AnimeGroup{
					Letter:    letter,
					AnimeList: animeList,
				})
			}
		})
	}

	return animeGroups, nil
}
